### PR TITLE
(GH-14) Add test cases for TFS calls

### LINKED ISSUE #14: Add test cases for TFS calls
We should add test cases for the core functionality of this addin using a mock of `GitHttpClient`

Original issue: https://github.com/cake-contrib/Cake.Issues.PullRequests.Tfs/issues/3

### PR DESCRIPTION
**Pull request by yansklyarenko**
@pascalberger This is the very first draft to serve as a proof of concept for the idea. It contains just one ugly ctor test using the mocked `GitHttpClient`. I had to make the interface of the factory `public` since it complained when being used from the `public` ctor of the `TfsPullRequest`. 

In general, does it resemble what you have in mind? I would appreciate your review. 

Fixes #14.

### CODE REVIEW

File: src/Cake.Tfs/PullRequest/GitClient/IGitClientFactory.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    /// The interface for a Git client factory
+    /// </summary>
+    public interface IGitClientFactory
+    {
+        /// <summary>
+        /// Creates the instance of the <see cref="GitHttpClient"/> class.
+        /// </summary>
+        /// <param name="collectionUrl">The URL of the TFS/Azure DevOps team project collection.</param>
+        /// <param name="credentials">The credentials to connect to TFS/Azure DevOps.</param>
+        /// <returns>The insntace of <see cref="GitHttpClient"/> class.</returns>
```
Comment (pascalberger) [inline src/Cake.Tfs/PullRequest/GitClient/IGitClientFactory.cs]: Typo `insntace`

File: src/Cake.Tfs/PullRequest/GitClient/IGitClientFactory.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        GitHttpClient CreateGitClient(Uri collectionUrl, ITfsCredentials credentials);
+
+        /// <summary>
+        /// Creates the instance of the <see cref="GitHttpClient"/> class.
+        /// </summary>
+        /// <param name="collectionUrl">The URL of the TFS/Azure DevOps team project collection.</param>
+        /// <param name="credentials">The credentials to connect to TFS/Azure DevOps.</param>
+        /// <param name="identity">Returns identity which is authorized.</param>
+        /// <returns>The insntace of <see cref="GitHttpClient"/> class.</returns>
```
Comment (pascalberger) [inline src/Cake.Tfs/PullRequest/GitClient/IGitClientFactory.cs]: Typo `insntace`

File: src/Cake.Tfs/PullRequest/GitClient/IGitClientFactory.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+﻿namespace Cake.Tfs.PullRequest.GitClient
```
Comment (pascalberger) [inline src/Cake.Tfs/PullRequest/GitClient/IGitClientFactory.cs]: I would suggest to move the factory to the `Cake.Tfs` namespace, as it might be used in the future for other stuff beside pull requests

File: src/Cake.Tfs.Tests/PullRequest/TfsPullRequestTests.cs [hunk 4]
```diff
@@ -32,13 +32,30 @@ public void Should_Throw_If_Settings_Are_Null()
...
+                var log = new FakeLog();
+                var settings = new TfsPullRequestSettings(
+                    new Uri("http://MyServer/tfs/MyCollection/MyTeamProject/_git/MyRepoName"), 42, new TfsNtlmCredentials());
+                var gitClientFactory = new FakeGitClientFactory();
+
+                // When
+                var pullRequest = new TfsPullRequest(log, settings, gitClientFactory);
+
+                // Then
+                Assert.Equal(42, pullRequest.PullRequestId);
```
Comment (pascalberger) [inline src/Cake.Tfs.Tests/PullRequest/TfsPullRequestTests.cs]: I would use Shouldly for assertion checking, like in other test cases:

```
pullRequest.PullRequestId.ShouldBe(42);
```

File: src/Cake.Tfs.Tests/PullRequest/TfsPullRequestTests.cs [hunk 5]
```diff
@@ -32,13 +32,30 @@ public void Should_Throw_If_Settings_Are_Null()
...
             }
+
+            [Fact]
+            public void Should_Return_Valid_Pull_Request()
+            {
+                // Given
+                var log = new FakeLog();
+                var settings = new TfsPullRequestSettings(
+                    new Uri("http://MyServer/tfs/MyCollection/MyTeamProject/_git/MyRepoName"), 42, new TfsNtlmCredentials());
+                var gitClientFactory = new FakeGitClientFactory();
```
Comment (pascalberger) [inline src/Cake.Tfs.Tests/PullRequest/TfsPullRequestTests.cs]: To simplify writing of test cases we could create a fixture class, which does all this bootstrapping code to some default values, which can be overwriten as required by the specific test case (from the principle like done [here](https://github.com/cake-contrib/Cake.Issues.PullRequests/blob/develop/src/Cake.Issues.PullRequests.Tests/PullRequestsFixture.cs))

**Comment by yansklyarenko** [comment]
@pascalberger Thanks for your review! I've corrected the small issues you outlined and followed your advice with the fixture class. Hope I got it right. At this point, I would say that ctor is pretty much covered. Could you please take a look? As soon as we are done with the ctor tests, I'll proceed with `Vote` and `SetStatus` methods.

**Comment by pascalberger** [comment]
@yansklyarenko Looks good to me so far 👍

**Comment by yansklyarenko** [comment]
@pascalberger Ok, I have created the tests for `Vote` and `SetStatus`. At this point, I think I have covered what was planned. Please, review.

I have several questions regarding the design overall:

- The methods `Vote` and `SetStatus` don't return a value, that's why I couldn't figure out how to test the happy path. I have left the comments in the appropriate places in the tests. I tried to cover the corner cases of these methods, that is, the possible exceptions which could be thrown. Do these methods return `void` on purpose? As far as I can see, when called from the Cake alias, the return value is never used anyway. Do you think it makes sense to let them return `true`/`false`, for example? Or the object of status being set or the vote being casted?

- It looks like the hierarchy of the fake git client factories has gone too far. The main one sets up the normal calls of the `GitHttpClient` methods, the `Null...` one returns `null` for those methods called from the ctor, and the `NullForMethods...` factory sets up the normal call for the method called from ctor, but returns `null` for everything else. This factory turned out to be useful in just couple of tests, but led to some code duplication. That's why I'm a bit concerned whether it is required at all... What do you think?

- This code of ctor is never reached:
`                else
                {
                    throw new ArgumentOutOfRangeException(
                        nameof(settings),
                        "Either PullRequestId or SourceBranch needs to be set");
                }
` 
The Settigns class is designed the way not to create an object in case both PR ID and source branch are not set. At least, I failed to get to that part of the flow in the tests. Should it stay?

**Comment by yansklyarenko** [comment]
@pascalberger The last commit addresses the point number 2 above. Now it looks somewhat prettier to me.

File: src/Cake.Tfs/PullRequest/TfsPullRequest.cs [hunk 6]
```diff
@@ -196,6 +198,26 @@ public int CodeReviewId
...
+            get
+            {
+                if (!this.ValidatePullRequest())
+                {
+                    return string.Empty;
+                }
+
+                return this.pullRequest.SourceRefName;
+            }
+        }
```
Comment (pascalberger) [inline src/Cake.Tfs/PullRequest/TfsPullRequest.cs]: This is adds a separate enhancement. To have this appear in the release notes properly I've created https://github.com/cake-contrib/Cake.Tfs/issues/36

File: src/Cake.Tfs.Tests/ExceptionAssertExtensions.cs [hunk 7]
```diff
@@ -49,5 +50,32 @@ public static void IsInvalidOperationException(this Exception exception)
...
+        public static void IsTfsPullRequestNotFoundException(this Exception exception)
+        {
+            Assert.IsType<TfsPullRequestNotFoundException>(exception);
+        }
+
+        /// <summary>
+        /// Checks if an exception is of type <see cref="NullReferenceException"/>.
+        /// </summary>
+        /// <param name="exception">Exception to check.</param>
+        public static void IsNullReferenceException(this Exception exception)
```
[the remark this answers is not in the document]
Comment (yansklyarenko) [reply src/Cake.Tfs.Tests/ExceptionAssertExtensions.cs]: Fixed: the null ref checks were removed.

File: src/Cake.Tfs/PullRequest/TfsPullRequest.cs (hunk 6 above)
Comment (yansklyarenko) [reply src/Cake.Tfs/PullRequest/TfsPullRequest.cs]: @pascalberger Do you mean this should be pulled out from this branch and merged as a separate PR?

Comment (pascalberger) [reply src/Cake.Tfs/PullRequest/TfsPullRequest.cs]: A separate PR would be a cleaner approach. If you want you can move it to a separate PR, but I'm OK to keep it here. The important thing is that an issue exists for it so that it appears in the release notes later.

Comment (yansklyarenko) [reply src/Cake.Tfs/PullRequest/TfsPullRequest.cs]: @pascalberger Makes sense. I've pulled out that part. As soon as this PR is merged, I'll issue a new one with just that change, referencing the new issue you created.

**Comment by pascalberger** [comment]
@yansklyarenko Your changes have been merged, thanks for your contribution 👍
